File: apps/mobile/ios/Runner/Native/Voice/VoiceHostApiImpl.swift
import Foundation

final class VoiceHostApiImpl: NSObject, VoiceHostApi {
  let manager: VoiceManager
  private let flutterApi: VoiceFlutterApi

  init(manager: VoiceManager, flutterApi: VoiceFlutterApi) {
    self.manager = manager
    self.flutterApi = flutterApi
    super.init()
    manager.onCommand = { [weak self] cmd in
      self?.flutterApi.onWakeDetected(command: cmd == .start ? .start : .stop) { _ in }
    }
    manager.onStateChanged = { [weak self] state in
      self?.flutterApi.onListeningStateChanged(state: state) { _ in }
    }
  }

  func isAvailable(completion: @escaping (Result<Bool, Error>) -> Void) {
    manager.isAvailable { ok in completion(.success(ok)) }
  }

  func startListening() throws { manager.start() }
  func stopListening() throws { manager.stop() }
}

File: apps/mobile/android/app/src/main/kotlin/com/rarocamera/raro_mobile/voice/VoiceHostApiImpl.kt
package com.rarocamera.raro_mobile.voice

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.rarocamera.raro_mobile.generated.voice.VoiceFlutterApi
import com.rarocamera.raro_mobile.generated.voice.VoiceHostApi
import com.rarocamera.raro_mobile.generated.voice.VoiceListeningState
import com.rarocamera.raro_mobile.generated.voice.WakeCommand

class VoiceHostApiImpl(
  private val context: Context,
  private val flutterApi: VoiceFlutterApi,
) : VoiceHostApi {
  private val main = Handler(Looper.getMainLooper())
  private val foreground = ForegroundVoiceRecognizer(
    context,
    onCommand = { cmd -> emitCommand(cmd) },
    onState = { state -> emitState(state) },
  )

  var wantsListening: Boolean = false
    private set

  init {
    VoiceBackgroundService.commandListener = { cmd -> emitCommand(cmd) }
  }

  override fun isAvailable(callback: (Result<Boolean>) -> Unit) {
    callback(Result.success(foreground.isAvailable()))
  }

  override fun startListening() {
    wantsListening = true
    main.post { foreground.start() }
  }

  override fun stopListening() {
    wantsListening = false
    main.post { foreground.stop() }
  }

  fun moveToBackground() {
    if (!wantsListening) return
    main.post {
      foreground.stop()
      VoiceBackgroundService.start(context)
    }
  }

  fun moveToForeground() {
    if (!wantsListening) return
    main.post {
      VoiceBackgroundService.stop(context)
      foreground.start()
    }
  }

  private fun emitCommand(cmd: WakeCommand) {
    main.post { flutterApi.onWakeDetected(cmd) {} }
  }

  private fun emitState(state: VoiceListeningState) {
    main.post { flutterApi.onListeningStateChanged(state) {} }
  }
}
